src/Interpreters/InterpreterBackupQuery.cpp

```diff
@@ -36,15 +38,18 @@ namespace
 
 BlockIO InterpreterBackupQuery::execute()
 {
+    const ASTBackupQuery & backup_query = query_ptr->as<const ASTBackupQuery &>();
     auto & backups_worker = context->getBackupsWorker();
-    auto id = backups_worker.start(query_ptr, context);
 
-    auto info = backups_worker.getInfo(id);
-    if (info.exception)
-        std::rethrow_exception(info.exception);
+    auto [id, status] = backups_worker.start(query_ptr, context);
+
+    /// Wait if it's a synchronous operation.
+    bool async = BackupSettings::isAsync(backup_query);
+    if (!async)
+        status = backups_worker.wait(id);
```

Comment:
To simplify the code any backups or restores are now run asynchronously inside the implementation. Synchronous backups just waits for completion. It's simpler this way than to always take into account which thread to use inside `BackupsWorker`.